packages/commonwealth/server/migrations/20231002202025-cosmos-hex.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+const getHex = async (address) => {
+  const { toHex, fromBech32 } = await import('@cosmjs/encoding');
+  const encodedData = fromBech32(address).data;
+  const addressHex = toHex(encodedData);
+  return addressHex;
+};
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.sequelize.transaction(async (t) => {
+      await queryInterface.sequelize.query(
+        `
+        ALTER TABLE "Addresses" ADD COLUMN "hex" VARCHAR(64) NULL;
+        `,
+        { raw: true, transaction: t }
+      );
+
+      // get all cosmos addresses and assign a hex
+      const [addresses] = await queryInterface.sequelize.query(
+        `
+        SELECT id, address
+        FROM "Addresses"
+        WHERE "wallet_id" = 'keplr';
+        `,
+        { transaction: t }
+      );
+
+      for (const address of addresses) {
+        try {
+          const hex = await getHex(address.address);
+          await queryInterface.sequelize.query(
+            `
+          UPDATE "Addresses"
+          SET hex = '${hex}'
+          WHERE id = ${address.id};
+          `,
+            { transaction: t }
+          );
+        } catch (e) {
+          console.log(
+            `Error getting hex for ${address.address}. Not updating.`
+          );
+        }
+      }
+
+      // then consolidate similar hexes into user_id with latest last_active
+      const [hexes] = await queryInterface.sequelize.query(
+        `
+        SELECT hex, last_active
+        FROM "Addresses"
+        WHERE hex IS NOT NULL AND last_active IS NOT NULL
+        GROUP BY hex, last_active;
+        `,
+        { transaction: t }
+      );
+
+      for (const hex of hexes) {
+        const [address] = await queryInterface.sequelize.query(
+          `
+          SELECT id, user_id, profile_id
+          FROM "Addresses"
+          WHERE hex = '${hex.hex}'
+          ORDER BY last_active DESC
+          LIMIT 1;
+          `,
+          { transaction: t }
+        );
+
+        await queryInterface.sequelize.query(
+          `
+          UPDATE "Addresses"
+          SET user_id = ${address[0].user_id}, profile_id = ${address[0].profile_id}
+          WHERE hex = '${hex.hex}';
+          `,
+          { transaction: t }
+        );
+      }
+    });
+  },
+
+  down: async (queryInterface, Sequelize) => {
+    await queryInterface.sequelize.transaction(async (t) => {
```

Comment:
@timolegros As you can see, above we alter a lot of data. Changing `user_id`s and `profile_id`s. Can you imagine a way to undo these changes in the `down` function? Or do we rely on a backup DB in this type of case?

Reply:
If the `up` function is removing data/record relationships then those can't be recovered in the `down` function. Ideally, for changes like this, we use the [expand-contract pattern][1] which essentially just means that the old data still exists (we don't overwrite it) and simply drop it when the changes are confirmed.

That said, we have rarely ever done this since as far as I can remember we haven't needed to use the `down` function so it might be sufficient to say that we can use a database dump in the absolute worst-case scenario. While we really should get into a habit of writing reversible migrations, this takes a lot of time and isn't a priority at this time. My general rule of thumb up to this point has been I make a migration reversible if it takes me a reasonable amount of time (1-2 points but no more than that).

IMO it should be fine to leave `down` empty. I literally just did this [here][2].

I would like to note though that your approach to the migration might cause issues in production. You are performing thousands upon thousands of individual `UPDATE` queries within a single transaction (using for loop) which can lead to locking/performance issues. Let's hop on a call tomorrow and discuss query alternatives + explore if we can easily apply the expand-contract pattern.

[1]: https://medium.com/continuousdelivery/expand-contract-pattern-and-continuous-delivery-of-databases-4cfa00c23d2e
[2]: https://github.com/hicommonwealth/commonwealth/pull/5207/files#diff-d5cfbbb1b156adc70989778fe827cc2d04c497d4925685f6071f2d6bea4b208aR35

Reply:
Thanks. I'll read up on expand-contract.

It is a massive migration, and I'm open to suggestions. Maybe we can break it into separate transactions.